pkg/kallsyms/kallsyms_test.go

```diff
@@ -62,21 +54,178 @@ func TestCustomKAllSyms(t *testing.T) {
+			expectKmod: "",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.symbol, func(t *testing.T) {
+			r := strings.NewReader(kAllSymsStr)
+			addr, kmod, err := kernelSymbolAddressFromReader(r, test.symbol)
+
+			if test.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+			require.Equal(t, test.expectAddr, addr, "addr")
+			require.Equal(t, test.expectKmod, kmod, "kmod")
+		})
+	}
+}
+
+func TestRealKAllSymsParsing(t *testing.T) {
+	path := "/proc/kallsyms"
 	utilstest.RequireRoot(t)
-	utilstest.RequireFileContains(t, "/proc/kallsyms", "bpf_prog_fops")
-	utilstest.RequireFileContains(t, "/proc/kallsyms", "socket_file_ops")
-
-	kAllSyms, err := NewKAllSyms()
-	require.Nil(t, err, "NewKAllSyms failed: %v", err)
-	require.True(t, kAllSyms.SymbolExists("bpf_prog_fops"),
-		"SymbolExists should have found bpf_prog_fops")
-	require.False(t, kAllSyms.SymbolExists("abcde_bad_name"),
-		"SymbolExists should not have found abcde_bad_name")
-
-	addr, ok := kAllSyms.symbolsMap["bpf_prog_fops"]
-	require.True(t, ok, "bpf_prog_fops not found in symbolsMap")
-	// /proc/kallsyms contains the address 0 without CAP_SYSLOG.
-	// Since we use RequireRoot, it should not happen.
-	require.NotEqual(t, 0, addr, "bpf_prog_fops has a zero address")
+	utilstest.RequireFileContains(t, path, "bpf_prog_fops")
+	utilstest.RequireFileContains(t, path, "socket_file_ops")
+
+	type testT struct {
+		name           string
+		symbol         string
+		expectedKmod   string
+		expectedExists bool
+	}
+
+	tests := []testT{
+		{
+			name:           "simple_symbol1",
+			symbol:         "bpf_prog_fops",
+			expectedKmod:   "",
+			expectedExists: true,
+		},
+		{
+			name:           "simple_symbol2",
+			symbol:         "socket_file_ops",
+			expectedKmod:   "",
+			expectedExists: true,
+		},
+		{
+			name:           "symbol_from_veth_kmod",
+			symbol:         "veth_init",
+			expectedKmod:   "veth",
+			expectedExists: true,
+		},
+		{
+			name:           "symbol_from_netem_kmod",
+			symbol:         "netem_module_init",
+			expectedKmod:   "sch_netem",
+			expectedExists: true,
+		},
+		{
+			name:           "symbol_from_overlay_kmod",
+			symbol:         "ovl_inode_init",
+			expectedKmod:   "overlay",
+			expectedExists: true,
+		},
+		{
+			name:           "nonexistent_symbol",
+			symbol:         "abcde_bad_name",
+			expectedKmod:   "",
+			expectedExists: false,
+		},
+	}
+
+	kallsymsBytes, err := os.ReadFile(path)
+	require.Nil(t, err, "Failed to read %q: %s", path, err)
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.expectedExists &&
+				test.expectedKmod != "" &&
+				!strings.Contains(string(kallsymsBytes), test.expectedKmod) {
+				t.Skipf("Test requires kernel module %q", test.expectedKmod)
+			}
+
+			if SymbolExists(test.symbol) != test.expectedExists {
+				t.Errorf("SymbolExists(%s) = %v, want %v",
+					test.symbol,
+					!test.expectedExists,
+					test.expectedExists)
+			}
```

Comment:
It's also possible to use require.Equal for this one, If you don't want the test to fail immediately, you can use the assert package: https://pkg.go.dev/github.com/stretchr/testify/assert#Equal